### PR TITLE
Docs: Added Segment Definitions

### PR DESCRIPTION
**Pull request by MattBuckley-TfN**
# Changes

Added segment definitions to documentation and updated API documentation templates to new ones from [cookiecutter-caf](https://github.com/Transport-for-the-North/cookiecutter-caf).

New [User Guide > Data Definitions page](https://cafbase--136.org.readthedocs.build/en/136/definitions.html)

# Tasks

- [x] Have unittests been added (testing individual functions and their edge cases) - NA
- [x] Has documentation (including user guide) been updated - Yes
- [x] Have new dependencies been added (or changed) in relevant `requirements.txt` - No
- [x] Code linting, tests and other workflows pass

### CODE REVIEW

File: docs/source/_templates/autosummary/class.rst [hunk 1]
```diff
@@ -10,7 +26,9 @@
 
    .. autosummary::
    {% for item in attributes %}
+      {%- if include_inherited_attributes|default(false) or item not in inherited_members %}
       ~{{ name }}.{{ item }}
+      {%- endif -%}
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply docs/source/_templates/autosummary/class.rst:31]: inherited_members comes from autosummary so should always be there

File: docs/source/_templates/autosummary/class.rst [hunk 2]
```diff
@@ -22,7 +40,9 @@
    .. autosummary::
       :toctree:
    {% for item in methods %}
+      {%- if include_inherited_methods|default(false) or item not in inherited_members %}
       ~{{ name }}.{{ item }}
+      {%- endif -%}
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply docs/source/_templates/autosummary/class.rst:45]: inherited_members comes from autosummary so should always be there

File: docs/source/_templates/autosummary/class.rst [hunk 3]
```diff
@@ -32,7 +52,14 @@
    {% if attributes %}
    .. rubric:: Attributes Documentation
    {% for item in attributes %}
+   {%- if include_inherited_attributes|default(false) or item not in inherited_members %}
    .. autoattribute:: {{ name }}.{{ item }}
+   {%- endif -%}
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply docs/source/_templates/autosummary/class.rst:57]: inherited_members comes from autosummary so should always be there

File: docs/source/conf.py [hunk 4]
```diff
@@ -208,10 +227,38 @@ def linkcode_resolve(domain: str, info: dict) -> str | None:
     except ValueError:
         return None
 
+    return f"{get_github_url()}/{filepath}"
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply docs/source/conf.py:232]: Seems to work fine for now but worth bearing in mind for future if some issues do crop up

**Comment by MattBuckley-TfN** [comment]
Ignoring the MyPy issues (in Lint) because they're coming from main

**Review by BenTaylor-TfN — DISMISSED** [review]

**Review by BenTaylor-TfN — APPROVED** [review]
